### DISCUSSION TITLE
Connect Prisma to Google Cloud Sql

### DISCUSSION THREAD

**Discussion by AminAppDevs** [prisma/prisma#12854] [Q&A]
I am tring to connect nestjs app using prisma to google cloud sql (postgres) - app on google app engine.
Is there any method to that?
any solution i found on the internet not work,
please help

**Comment by austincrim** [comment]
Hey @AminAppDevs 👋 ,

What kind of trouble are you running into? Any specific errors?

**Comment by janpio** [comment]
We have permanent tests with Google Cloud SQL, PostgreSQL, so this should definitely work: https://github.com/prisma/ecosystem-tests/tree/dev/databases/gcp-postgresql-ssl Please reply to @austincrim with more information what is actually going wrong for you and we might be able to help.

**Reply by gui8515** [reply]
helped me a lot!

**Reply by Tom-DK** [reply]
I think this method only works for a local connection of the Prisma Client to Cloud SQL, as it requires IP whitelisting. You don't want to whitelist IPs of a CI/CD provider...

Anyone that knows an internal workaround within GCP?

**Comment by mcabilo** [comment]
I always get the error P1001 from prisma saying that it cannot reach the database server.
I used the following format for url: 
postgresq://<user>:<password>@<public_ip>:5432/<db_name>?socket=/cloudsql/<project_id>:<region>:<db_id>

**Reply by dhruvvbhavsar** [reply]
Hey. facing the same error. how did you manage to solve it?

**Comment by baotpham** [comment]
this is how I did it:
```
postgresql://postgres:<password>@<public_ip_address>:<port>/postgres?host=/cloudsql/<project_id>:<connection_string>
```

**Reply by Joebayld** [reply]
This was helpful for me. It's worth noting that you don't have to use the public ip and port, you can just use localhost as the ip is ignored.

**Reply by bryhasagithub** [reply]
<3

**Comment by CarltonK** [comment]
I'm using NestJS + Prisma on Google Cloud Run (v1)
As a workaround, I have to download the cloud SQL binary in the docker image

```
RUN wget https://storage.googleapis.com/cloud-sql-connectors/cloud-sql-proxy/v2.13.0/cloud-sql-proxy.linux.amd64 -O cloud-sql-proxy
RUN chmod +x cloud-sql-proxy

# Create the /cloudsql directory
RUN mkdir -p /cloudsql && chown node:node /cloudsql

ARG SQL_HOST
ENV SQL_HOST=${SQL_HOST}

RUN chmod +x /app/entrypoint.sh

ENTRYPOINT ["/app/entrypoint.sh"]
```

My entrypoint script looks like this
```
#!/bin/sh
# Start the Cloud SQL Proxy in the background
./cloud-sql-proxy --auto-iam-authn ${SQL_HOST} --unix-socket /cloudsql &

# Start the NestJS server in the background
npm run start

wait

```

My connection string looks something like
`postgresql://SA_EMAIL@localhost/DB_NAME?connection_limit=1&host=/cloudsql/INSTANCE_CONNECTION_NAME`


I'm still trying to find a way in which I don't have to download the binary within my container and instead rely on the Cloud Sql Connections. I still have the option to switch to using a Cloud Sql Language Connector which works really well for cloud functions as seen below
```
private async connectToDatabase(): Promise<any> {
        try {
            if (this.dbConn) {
                return; // Already connected
            }

            let config: Knex.Config;

            if (process.env.NODE_ENV === 'local') {
                config = {
                    client: 'pg',
                    connection: {
                        host: `172.35.0.2`,
                        port: 5432,
                        user: 'user',
                        database: 'db',
                        password: 'password',
                    },
                    pool: { min: 0, max: 5 },
                };
            } else {
                const databaseSecrets = await this.secretManager.getSecret('database');
                const { workerUser, database, connectionName } = databaseSecrets;

                const connector = new Connector();
                const clientOpts = await connector.getOptions({
                    instanceConnectionName: connectionName,
                    authType: AuthTypes.IAM,
                });

                config = {
                    client: 'pg',
                    connection: { ...clientOpts, user: workerUser, database },
                    pool: { min: 0, max: 5 },
                };
            }
            
            this.dbConn = knex(config);
        } catch (error) {
            this.logger.error(`${error}`);
        }
    }
```
If anyone can help this would be appreciated.

**Comment by bryhasagithub** [comment]
any luck on this?

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!

**Comment by ggorge-me** [comment]
This has worked for me: `postgresql://user:pass@localhost/postgres?host=/cloudsql/gcp-project:gcp-region:cloudsql-db/`. Please note the trailing slash, because without it does not work!

**Comment by ernest0vm** [comment]
i just create a VPC network > add a subnet (with `Private Google Access` to `On`) > add Private Services Access > add VPC network peering
into the cloud run add the VPC network created (route only requests to private IPs to the VPC)
into the cloud sql add the private IP with the VPC created (it returns an internal/private IP)
set the `DATABASE_URL` value using `postgresql://user:password@<PRIVATE_IP>:5432/my-db`

**Comment by kdawgwilk** [comment]
An example of IAM auth using the cloud-sql-proxy nodejs sdk would very helpful
https://github.com/GoogleCloudPlatform/cloud-sql-nodejs-connector
